cli/cmd/cluster/certificate-rotator.go

```diff
@@ -0,0 +1,229 @@
+	if err != nil {
+		return fmt.Errorf("reading Kubernetes CA certificate from Terraform output: %w", err)
+	}
+
+	crConfig := certificateRotatorConfig{
+		clientSet:            cs,
+		newCACert:            newCACert,
+		logger:               contextLogger,
+		daemonSetsToRestart:  c.platform.Meta().DaemonSets,
+		deploymentsToRestart: c.platform.Meta().Deployments,
+	}
+
+	cr := certificateRotator{
+		config: crConfig,
+	}
+	if cr.validate() != nil {
+		return fmt.Errorf("preparing certificate rotator: %w", err)
+	}
+
+	// rotate() restarts control plane deployments and daemonsets again. We
+	// need to it again because the first rollout is partial.
+	if err := cr.rotate(); err != nil {
+		return fmt.Errorf("rotating certificates: %w", err)
+	}
+
+	return nil
+}
+
+func (cr *certificateRotator) validate() error {
+	if cr.config.clientSet == nil {
+		return fmt.Errorf("clientSet can't be nil")
+	}
+
+	if cr.config.newCACert == "" {
+		return fmt.Errorf("new CA certificate can't be empty")
+	}
+
+	return nil
+}
+
+// rotate will wait for service accounts to be signed by the new CA and restart all system
+// DaemonSets and Deployments using the CA certificate.
+func (cr *certificateRotator) rotate() error {
+	ctx, cancel := context.WithDeadline(context.Background(), time.Now().Add(k8sutil.RolloutTimeout))
+	defer cancel()
+
+	cr.config.logger.Printf("Waiting for all service account tokens on the cluster to be updated...")
+
+	if err := cr.waitForUpdatedServiceAccountTokens(ctx); err != nil {
+		return fmt.Errorf("waiting for all service account tokens to be updated: %w", err)
+	}
+
+	cr.config.logger.Printf("All service account tokens has been updated with new Kubernetes CA certificate")
+
+	for _, daemonSet := range cr.config.daemonSetsToRestart {
+		cr.config.logger.Printf("Restarting DaemonSet %s/%s to pick up new Kubernetes CA Certificate",
+			daemonSet.Namespace, daemonSet.Name)
+
+		if err := k8sutil.RolloutDaemonSet(cr.config.clientSet.AppsV1().DaemonSets(daemonSet.Namespace),
+			daemonSet.Name); err != nil {
+			return fmt.Errorf("restarting DaemonSet %s/%s: %w", daemonSet.Namespace, daemonSet.Name, err)
+		}
+	}
+
+	for _, deployment := range cr.config.deploymentsToRestart {
+		cr.config.logger.Printf("Restarting Deployment %s/%s to pick up new Kubernetes CA Certificate",
+			deployment.Namespace, deployment.Name)
+
+		if err := k8sutil.RolloutDeployment(cr.config.clientSet.AppsV1().Deployments(deployment.Namespace),
+			deployment.Name); err != nil {
+			return fmt.Errorf("restarting Deployment %s/%s: %w", deployment.Namespace, deployment.Name, err)
+		}
+	}
+
+	return nil
+}
+
+func (cr *certificateRotator) waitForUpdatedServiceAccountTokens(ctx context.Context) error {
+	for {
+		select {
+		case <-ctx.Done():
+			return fmt.Errorf("context exceeded while checking if all service account tokens include new CA certificate: %w", ctx.Err()) //nolint:lll
+		case <-time.After(time.Second):
```

Comment:
one second feels short to me...

Reply:
We only poll every second, not timeout after it. What is wrong with it?